src/librustc_mir/hair/pattern/check_match.rs

```diff
@@ -146,17 +143,15 @@ impl<'tcx> MatchVisitor<'_, 'tcx> {
             }
 
             // Third, perform some lints.
-            for pat in &arm.pats {
-                check_for_bindings_named_same_as_variants(self, pat);
-            }
+            check_for_bindings_named_same_as_variants(self, &arm.pat);
         }
 
         let module = self.tcx.hir().get_module_parent(scrut.hir_id);
         MatchCheckCtxt::create_and_enter(self.tcx, self.param_env, module, |ref mut cx| {
             let mut have_errors = false;
 
             let inlined_arms : Vec<(Vec<_>, _)> = arms.iter().map(|arm| (
-                arm.pats.iter().map(|pat| {
+                arm.top_pats_hack().iter().map(|pat| {
```

Comment:
@dlrobertson You might want to extract the exhaustiveness checking changes out from https://github.com/rust-lang/rust/pull/63688 to a dedicated PR and link up with the change here. That might allow us to parallelize adding tests for exhaustiveness / liveness stuff.